pkg/didcomm/protocol/issuecredential/service.go

```diff
@@ -0,0 +1,417 @@
+	if canTriggerActionEvents(msg) {
+		aEvent <- s.newDIDCommActionMsg(md)
+
+		return "", nil
+	}
+
+	// if no action event is triggered, continue the execution
+	return "", s.handle(md)
+}
+
+// HandleOutbound handles outbound message (issuecredential protocol)
+func (s *Service) HandleOutbound(msg service.DIDCommMsg, myDID, theirDID string) (string, error) {
+	md, err := s.doHandle(msg, true)
+	if err != nil {
+		return "", fmt.Errorf("doHandle: %w", err)
+	}
+
+	// sets outbound payload
+	md.MyDID = myDID
+	md.TheirDID = theirDID
+
+	return "", s.handle(md)
+}
+
+func (s *Service) doHandle(msg service.DIDCommMsg, outbound bool) (*metaData, error) {
+	piID, err := getPIID(msg)
+	if err != nil {
+		return nil, fmt.Errorf("piID: %w", err)
+	}
+
+	stateName, err := s.currentStateName(piID)
+	if err != nil {
+		return nil, fmt.Errorf("currentStateName: %w", err)
+	}
+
+	current := stateFromName(stateName)
+
+	next, err := nextState(msg, outbound)
+	if err != nil {
+		return nil, fmt.Errorf("nextState: %w", err)
+	}
+
+	if !current.CanTransitionTo(next) {
+		return nil, fmt.Errorf("invalid state transition: %s -> %s", current.Name(), next.Name())
+	}
+
+	return &metaData{
+		transitionalPayload: transitionalPayload{
+			StateName: next.Name(),
+			Msg:       msg.(service.DIDCommMsgMap),
+			PIID:      piID,
+		},
+		state:    next,
+		msgClone: msg.Clone(),
+	}, nil
+}
+
+// startInternalListener listens to messages in go channel for callback messages from clients.
+func (s *Service) startInternalListener() {
+	for msg := range s.callbacks {
+		// if no error do handle
+		if msg.err == nil {
+			msg.err = s.handle(msg)
+		}
+
+		// no error - continue
+		if msg.err == nil {
+			continue
+		}
+
+		msg.state = &abandoning{}
+
+		if err := s.handle(msg); err != nil {
+			logger.Errorf("listener handle: %s", err)
+		}
+	}
+}
+
+func isNoOp(s state) bool {
+	_, ok := s.(*noOp)
+	return ok
+}
+
+func (s *Service) handle(md *metaData) error {
+	var (
+		current   = md.state
+		actions   []stateAction
+		stateName string
+	)
+
+	for !isNoOp(current) {
+		stateName = current.Name()
+
+		next, action, err := s.execute(current, md)
+		if err != nil {
+			return fmt.Errorf("execute: %w", err)
+		}
+
+		actions = append(actions, action)
+
+		if !isNoOp(next) && !current.CanTransitionTo(next) {
+			return fmt.Errorf("invalid state transition: %s --> %s", current.Name(), next.Name())
+		}
+
+		current = next
+	}
+
+	if err := s.saveStateName(md.PIID, stateName); err != nil {
+		return fmt.Errorf("failed to persist state %s: %w", stateName, err)
+	}
+
+	for _, action := range actions {
+		if err := action(s.messenger); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func getPIID(msg service.DIDCommMsg) (string, error) {
+	if pthID := msg.ParentThreadID(); pthID != "" {
+		return pthID, nil
+	}
+
+	thID, err := msg.ThreadID()
+	if errors.Is(err, service.ErrThreadIDNotFound) {
+		// TODO: create a setter function e.g SetID
```

Comment:
shouldn't the messenger be setting the ID?

Reply:
When you send the first message you need to set this ID. I am going to add `SetID` function to the message interface.

Reply:
I still don't understand why the messenger can't set the ID.

It seems like we're setting the `@id` here and return it as the PIID for later use in the call to storage in `currentStateName`, where it will obviously not be found in this case.

So why not return an appropriate `ErrNotFound` from this method instead and earn two things:

* this logic won't have to be replicated in other protocols. Leave it up to the messenger to insert an `@id`
* save ourselves an unnecessary call to the storage

Reply:
Another option would be for `NewDIDCommMsgMap()` to set the `@id`.

Reply:
I forgot about an additional call to DB with not found result. Adding it to the construction function will not help, id and in that case, will not be empty. It makes sense to rely on the messenger. I am going to change this logic.

Reply:
@llorllale I have added `SetID` function. A call to DB is avoided now. I cannot rely on the messenger because I need to save data before action will be executed.